tests/integration/test_s3.py

```diff
@@ -121,3 +218,146 @@ def test_s3_multipart(self):
+            CreateBucketConfiguration={'LocationConstraint': cls.region})
+
+    def setUp(self):
+        self.files = FileCreator()
+
+    def tearDown(self):
+        self.files.remove_all()
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.client.delete_bucket(Bucket=cls.bucket_name)
+
+    def delete_object(self, key):
+        self.client.delete_object(
+            Bucket=self.bucket_name,
+            Key=key)
+
+    def object_exists(self, key):
+        self.client.head_object(Bucket=self.bucket_name,
+                                Key=key)
+        return True
+
+    def create_s3_transfer(self):
+        return boto3.s3.transfer.S3Transfer(self.client)
+
+    def test_upload_below_threshold(self):
+        transfer = self.create_s3_transfer()
+        filename = self.files.create_file_with_size(
+            'foo.txt', filesize=1024 * 1024)
+        transfer.upload_file(filename, self.bucket_name,
+                             'foo.txt')
+        self.addCleanup(self.delete_object, 'foo.txt')
+
+        self.assertTrue(self.object_exists('foo.txt'))
+
+    def test_upload_above_threshold(self):
+        transfer = self.create_s3_transfer()
+        filename = self.files.create_file_with_size(
+            '20mb.txt', filesize=20 * 1024 * 1024)
+        transfer.upload_file(filename, self.bucket_name,
+                             '20mb.txt')
+        self.addCleanup(self.delete_object, '20mb.txt')
+        self.assertTrue(self.object_exists('20mb.txt'))
+
+    def test_progress_callback_on_upload(self):
+        self.amount_seen = 0
+        lock = threading.Lock()
+
+        def progress_callback(amount):
+            with lock:
+                self.amount_seen += amount
+
+        transfer = self.create_s3_transfer()
+        filename = self.files.create_file_with_size(
+            '20mb.txt', filesize=20 * 1024 * 1024)
+        transfer.upload_file(filename, self.bucket_name,
+                             '20mb.txt', callback=progress_callback)
+        self.addCleanup(self.delete_object, '20mb.txt')
+
+        # The callback should have been called enough times such that
+        # the total amount of bytes we've seen (via the "amount"
+        # arg to the callback function) should be the size
+        # of the file we uploaded.
+        self.assertEqual(self.amount_seen, 20 * 1024 * 1024)
+
+    def test_can_send_extra_params_on_upload(self):
+        pass
+
+    def test_can_configure_threshold(self):
+        config = boto3.s3.transfer.TransferConfig(
+            multipart_threshold=6 * 1024 * 1024
+        )
+        transfer = boto3.s3.transfer.S3Transfer(self.client, config)
+        filename = self.files.create_file_with_size(
+            'foo.txt', filesize=8 * 1024 * 1024)
+        transfer.upload_file(filename, self.bucket_name,
+                             'foo.txt')
+        self.addCleanup(self.delete_object, 'foo.txt')
+
+        self.assertTrue(self.object_exists('foo.txt'))
```

Comment:
May want to assert that one of the two upload paths were taken based on the user-supplied threshold. Right now we only assert that the parameter can be passed in and a file exists after uploading. We don't know that the parameter was actually used.